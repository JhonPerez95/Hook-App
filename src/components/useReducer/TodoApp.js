import React, { useReducer } from 'react';

import todoReducer from './todoReducer';
import useForm from '../../hook/useForm';

import './style.css';

const initialState = [
  {
    id: new Date().getTime(),
    description: 'Aprender React',
    done: false,
  },
];

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, initialState);

  const [formState, handleInputChange, resetInput] = useForm({
    description: '',
  });
  const { description } = formState;

  const handleSunmit = (e) => {
    e.preventDefault();

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    const action = {
      type: 'ADD_TODO',
      payload: newTodo,
    };

    dispatch(action);
    resetInput();
  };

  return (
    <div>
      <h1>TodoApp: {todos.length}</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <ul className="list-group">
            {todos.map((item, i) => (
              <li
                key={item.id}
                className="list-group-item list-group-item-action"
              >
                <p className="text-center">
                  {i + 1}. {item.description}
                </p>
                <button className="btn btn-danger">Delete</button>
              </li>
            ))}
          </ul>
        </div>
        <div className="col-4">
          <h3>Add ToDo:</h3>
          <hr />
          <form onSubmit={handleSunmit}>
            <input
              type="text"
              className="form-control"
              name="description"
              placeholder="Comprar ..."
              autoComplete="off"
              value={description}
              onChange={handleInputChange}
            />
            <button
              className="btn btn-outline-primary mt-2 btn-block"
              type="submit"
            >
              Save ToDo
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
