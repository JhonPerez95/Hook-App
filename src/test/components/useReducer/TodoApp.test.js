import React from 'react';
import { shallow, mount } from 'enzyme';
import TodoApp from '../../../components/useReducer/TodoApp';
import { act } from '@testing-library/react';
import { demoTodos } from './demoTodos';

describe('Pruebas al component <TodoApp/>', () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn(() => {});

  test('Debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe Agregar un ToDo', () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find('TodoForm').prop('addTodo')(demoTodos[0]);
      wrapper.find('TodoForm').prop('addTodo')(demoTodos[1]);
    });
    expect(wrapper.find('h1').text().trim()).toBe('TodoApp: 2');
    expect(localStorage.setItem).toHaveBeenCalled();
  });

  test('Debe Eliminar un ToDo', () => {
    wrapper.find('TodoForm').prop('addTodo')(demoTodos[0]);
    wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

    expect(wrapper.find('h1').text().trim()).toBe('TodoApp: 0');
  });
});
