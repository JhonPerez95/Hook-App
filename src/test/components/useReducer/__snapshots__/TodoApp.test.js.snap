// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al component <TodoApp/> Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    TodoApp: 
    0
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-4"
    >
      <TodoForm
        addTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
